pkg/service/backup/worker_deduplicate.go

```diff
@@ -0,0 +1,232 @@
+// Copyright (C) 2024 ScyllaDB
+
+package backup
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"path"
+	"slices"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/scylladb/scylla-manager/v3/pkg/scyllaclient"
+	"github.com/scylladb/scylla-manager/v3/pkg/sstable"
+	"github.com/scylladb/scylla-manager/v3/pkg/util/parallel"
+)
+
+var (
+	// DeduplicateHostBeforeInterceptor is the function executed at the beginning of every deduplicate hots.
+	DeduplicateHostBeforeInterceptor = func() {}
+	// DeduplicateHostAfterInterceptor is a function executed after every deduplicate host invocation.
+	DeduplicateHostAfterInterceptor = func(dirs []SnapshotDir) {}
+)
```

Comment:
I don't really like having some exported global variables with no concurrent access control (not to mention that the second function allows for modifying snapshot dirs passed as function argument).

Couldn't we use the already existing client interceptors and wait for (and block) calls to `"POST": "/rclone/sync/movedir"`, so that we know when the deduplication stage is finished?
Couldn't we fetch progress with the usual `GetProgress` and get the size/deduplicated/uploaded from there?

This would also allow for not exporting the internals `snapshotDir` struct.

Reply:
using already existing interceptor is not option, it won't work here.
Will think about syncing (if it's needed).

This is not the perfect solution I agree, but should be fine. It's clear that these variables are for testing only.

Reply:
If I cancel the context from `/rclone/sync/movedir` then I'm doing it after deduplicate stage, so everything is deduplicated alteady.
To achieve the same effect as I have there, I would need to intercept RcloneSetBandwidthLimit.
I see no direct equivalent of after.

> (not to mention that the second function allows for modifying snapshot dirs passed as function argument).

It allows modifying snapshot dir, but I don't see it as a problem. Why do you think it's a problem ?

Reply:
But `DeduplicateHostAfterInterceptor` also happens after the deduplication is done.
The `DeduplicateHostBeforeInterceptor` could be replaced by `"POST": "/rclone/operations/deletefile"` with filtering for sstable file. Then you can use mutex to allow deduplication only on a single host, like in your current approach.

>>  (not to mention that the second function allows for modifying snapshot dirs passed as function argument).
> It allows modifying snapshot dir, but I don't see it as a problem. Why do you think it's a problem ?

Because it's unsafe to modify them there  - backup process assume they are exactly the same as the indexing stage created them. I know that we know that they shouldn't be modified, but we should try to write code which is robust even when somebody new looks at it.

So in general I'm in favor of refactoring it to the client interceptors, or to some more general  approach (e.g. allow for interceptors between any given stages), but I'm fine with it in the context of this PR, as long as those interceptors are not exported (they can be exported to tests only via `export_test.go`) , are marked as "use for testing only" in the comments, and they return a copy of snapshot dirs.
On the other hand, making a copy might be a pretty expensive operation, so maybe it would be better to just refactor function to take the already aggregated info and not snapshot dirs directly, like:
```go
deduplicateHostAfterTestInterceptor = func(deduplicated, uploaded, totalSize int64) {}
```

Reply:
OK, I found better approach.
Will hide it with build tags, so it's gonna be available for integration tests only.

Reply:
OK, it's refactored. 
All the hooks are hidden and available for the builds that includes `integration` or `all` tags.
`snapshotDirs` is unexported again.
Hooks belong to service and worker now.

Reply:
@karol-kokoszka but now `SetDeduplicateTestHooks` is exported and can be used in the backup pkg (not only in the tests)?
I think that using `export_test.go` could solve this problem, but I'm fine with merging this PR in its current state.

Reply:
`SetDeduplicateTestHooks` is not exported.
It's guarded with build tags

```
//go:build all || integration
// +build all integration
```

You can verify it by adding the following to backup/service.go:
```
	s.SetDeduplicateTestHooks(func() {}, func(skipped, uploaded, size int64) {})
```

and then try to compile `make build`

=> 
```
➜  scylla-manager git:(cheksum_comparison) ✗ make build
==> Building sctool
==> Building agent
==> Building scylla-manager
# github.com/scylladb/scylla-manager/v3/pkg/service/backup
pkg/service/backup/service.go:790:4: s.SetDeduplicateTestHooks undefined (type *Service has no field or method SetDeduplicateTestHooks)
make: *** [Makefile:201: build-server] Error 1

```